iwyu.cc

```diff
@@ -4123,25 +4125,26 @@ class IwyuAstConsumer
     if (CanIgnoreCurrentASTNode())
       return true;
     // TypedefType::getDecl() returns the place where the typedef is defined.
-    if (CanForwardDeclareType(current_ast_node())) {
-      ReportDeclForwardDeclareUse(CurrentLoc(), type->getDecl());
-    } else {
-      ReportDeclUse(CurrentLoc(), type->getDecl());
-    }
+    ReportDeclUse(CurrentLoc(), type->getDecl());
+    if (!CanForwardDeclareType(current_ast_node()))
+      ReportTypeUse(CurrentLoc(), type);
     return Base::VisitTypedefType(type);
   }
 
   bool VisitUsingType(clang::UsingType* type) {
     if (CanIgnoreCurrentASTNode())
       return true;
 
-    // UsingType is similar to TypedefType, so treat it the same.
```

Comment:
This makes me wonder why using-declarations do not have the same semantics as typedefs with regards to author intent. E.g.

```
// vended_aliases.h
class Foo;
typedef Foo foo_t;
using foo_u = Foo;

// provided_aliases.h
#include "bar.h"
typedef Bar bar_t;
using bar_u = Bar;

// main.cc
#include "vended_aliases.h"
#include "provided_aliases.h"

void test() {
    foo_t a;  // requires complete Foobar
    foo_u b;  // does not?   

    bar_t c; // provided by header, right?
    bar_u d; // not sure what the story is here
}
```

Reply:
(trying this out on master now)

Reply:
Surprisingly, all of the declarations in `test` are visited as `TypedefType` now and no `UsingType`s are seen. It's as if Clang has reverted to treating `UsingType` as discardable sugar.

Reply:
It's just a naming mess in C++. `clang::UsingType` is about `using ns::Type;`, not about `using Typedef = Type;`.

Reply:
:exploding_head:. Thanks. That makes sense, though the comparison to a typedef seems a little unfair now.